example/custom_authentication_cls.py

```diff
@@ -0,0 +1,62 @@
+from sanic import Sanic
+from sanic.response import json
+
+from sanic_jwt import Authentication, exceptions, Initialize
+
+
+class MyAuthentication(Authentication):
+    async def authenticate(self, request, *args, **kwargs):
+        username = request.json.get("username", None)
+        password = request.json.get("password", None)
+
+        if not username or not password:
+            raise exceptions.AuthenticationFailed(
+                "Missing username or password."
+            )
+
+        return {"user_id": 1}
+
+    async def store_refresh_token(
+        self, user_id, refresh_token, *args, **kwargs
+    ):
+        key = "refresh_token_{user_id}".format(user_id=user_id)
+        self.app.my_cache[key] = refresh_token
+
+    async def retrieve_refresh_token(self, user_id, *args, **kwargs):
+        key = "refresh_token_{user_id}".format(user_id=user_id)
+        token = self.app.my_cache.get(key, None)
+        return token
+
+    async def retrieve_user(self, request, payload, *args, **kwargs):
+        if payload:
+            user_id = payload.get("user_id", None)
+            return {"user_id": user_id}
+        else:
+            return None
+
+
+# elsewhere in the universe ...
+if __name__ == "__main__":
+    app = Sanic(__name__)
+    app.my_cache = {}
+
+    sanic_jwt = Initialize(
+        app, authentication_class=MyAuthentication, refresh_token_enabled=True
+    )
+
+    @app.route("/")
+    async def helloworld(request):
+        return json({"hello": "world"})
+
+    @app.route("/protected")
+    @sanic_jwt.protected()
```

Comment:
@vltr  What's your thought on this? Should the default recommendation be to use `@sanic_jwt.protected` or just `@protected`.

`@sanic_jwt.protected` clearly has the benefit of greater flexibility.
`@protected` has the benefit of simplicity since there is no need to pass around that instance. It is much easier to break an app into smaller modules and files and pull the decorator from a package than from an instance.

Reply:
I agree. For simple examples, and throughout **most** of the documentation, I agree let's stick to `@<decorator>`.

Maybe we also should move the one page examples into a subfolder and create a sample `complex-app`

```
./examples
├── complex-app
│   ├── bar
│   │   └── __init__.py
│   ├── foo
│   │   └── __init__.py
│   └── server.py
└── simple
    ├── basic.py
    ├── basic_with_certs.py
    ├── cbv.py
    ├── custom_endpoint.py
    ├── extended_payload.py
    ├── gettingstarted.py
    ├── main.py
    ├── on_blueprint.py
    ├── requirements.in
    ├── requirements.txt
    ├── scopes.py
    └── users.csv
```

Reply:
Yeah, this makes sense ... Even though some "complex app" examples we have are also snippets, perhaps with a good README.md inside every directory we can get things done.